tests/collections/common/test_ema.py

```diff
@@ -0,0 +1,400 @@
+            logger=False,
+            enable_model_summary=False,
+            accelerator='cpu',
+            devices=1,
+            callbacks=EMA(decay=0.999),
+        )
+        with pytest.raises(MisconfigurationException, match="Apex EMA Callback only works with CUDA"):
+            trainer.fit(model=model)
+
+    @mock.patch('nemo.collections.common.callbacks.ema.apex_available', False)
+    def test_ema_apex_unavailable(self):
+        with pytest.raises(MisconfigurationException, match="EMA requires Apex to be installed"):
+            EMA(decay=0.999)
+
+    @pytest.mark.unit
+    @pytest.mark.run_only_on('GPU')
+    @pytest.mark.skipif(not apex_available, reason="apex is not installed")
+    def test_ema_saved_state(self, tmpdir, caplog):
+        """Test to ensure that when we re-load the EMA callback, it loads the EMA weights correctly."""
+        temp_path = os.path.join(tmpdir, 'saved_state')
+
+        class TerminateCallback(Callback):
+            def on_train_epoch_end(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") -> None:
+                ema_callback = [x for x in trainer.callbacks if isinstance(x, EMA)][0]
+                self.saved_ema_weights = ema_callback._ema_model_weights
+                self.pl_module_weights = list(pl_module.state_dict().values())
+                raise SystemExit
+
+        model = ExampleModel()
+        terminate_callback = TerminateCallback()
+
+        trainer = Trainer(
+            max_epochs=2,
+            limit_val_batches=1,
+            limit_train_batches=16,
+            logger=False,
+            val_check_interval=0.5,
+            enable_checkpointing=False,
+            accelerator='gpu',
+            devices=1,
+            callbacks=[terminate_callback],
+        )
+        exp_manager(
+            trainer,
+            {
+                "ema": {"enable": True, "evaluate_ema_weights_instead": True},
+                "explicit_log_dir": str(temp_path),
+                "checkpoint_callback_params": {"filename": f"{{epoch}}-{{step}}"},
+            },
+        )
+        with pytest.raises(SystemExit):
+            trainer.fit(model=model)
+        resume_path = os.path.join(temp_path, 'checkpoints/epoch=0-step=8.ckpt')
+
+        model = ExampleModel()
+
+        class CheckStateCallback(Callback):
+            def on_train_start(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") -> None:
+                ema_callback = [x for x in trainer.callbacks if isinstance(x, EMA)][0]
+                weights = list(pl_module.state_dict().values())
+                for x, y in zip(weights, terminate_callback.pl_module_weights):
+                    assert torch.allclose(x.cpu(), y.cpu())
+                for x, y in zip(ema_callback._ema_model_weights, terminate_callback.saved_ema_weights):
+                    assert torch.allclose(x.cpu(), y.cpu())
+                assert ema_callback._cur_step == 8
+
+        trainer = Trainer(
```

Comment:
This test might be a bit expensive. Well let's measure it's runtime and if it's over 30 seconds reconsider the data set size

Reply:
All tests together in this file take around 30s total, since it's all using fake data the longest thing is read/write and CUDA initialisation!

Reply:
Hmm I suppose that's fine. We can't do much about cuda init.